File: lib/core/districts.dart
/// The 64 districts of Bangladesh, in their current official spellings.
///
/// District is what groups outbreak reports, so it has to be chosen from
/// a list rather than typed. Typed, one farmer writes "Bogra", another
/// "Bogura", a third "Bagura" — and a real outbreak split three ways
/// never reaches the alert threshold at all.
///
/// The backend keeps the same list in backend/districts.py. Change both
/// together.
library;

class District {
  final String name;
  final String division;

  const District(this.name, this.division);
}

const districts = <District>[
  // Dhaka
  District('Dhaka', 'Dhaka'),
  District('Faridpur', 'Dhaka'),
  District('Gazipur', 'Dhaka'),
  District('Gopalganj', 'Dhaka'),
  District('Kishoreganj', 'Dhaka'),
  District('Madaripur', 'Dhaka'),
  District('Manikganj', 'Dhaka'),
  District('Munshiganj', 'Dhaka'),
  District('Narayanganj', 'Dhaka'),
  District('Narsingdi', 'Dhaka'),
  District('Rajbari', 'Dhaka'),
  District('Shariatpur', 'Dhaka'),
  District('Tangail', 'Dhaka'),
  // Chattogram
  District('Bandarban', 'Chattogram'),
  District('Brahmanbaria', 'Chattogram'),
  District('Chandpur', 'Chattogram'),
  District('Chattogram', 'Chattogram'),
  District("Cox's Bazar", 'Chattogram'),
  District('Cumilla', 'Chattogram'),
  District('Feni', 'Chattogram'),
  District('Khagrachhari', 'Chattogram'),
  District('Lakshmipur', 'Chattogram'),
  District('Noakhali', 'Chattogram'),
  District('Rangamati', 'Chattogram'),
  // Rajshahi
  District('Bogura', 'Rajshahi'),
  District('Chapai Nawabganj', 'Rajshahi'),
  District('Joypurhat', 'Rajshahi'),
  District('Naogaon', 'Rajshahi'),
  District('Natore', 'Rajshahi'),
  District('Pabna', 'Rajshahi'),
  District('Rajshahi', 'Rajshahi'),
  District('Sirajganj', 'Rajshahi'),
  // Khulna
  District('Bagerhat', 'Khulna'),
  District('Chuadanga', 'Khulna'),
  District('Jashore', 'Khulna'),
  District('Jhenaidah', 'Khulna'),
  District('Khulna', 'Khulna'),
  District('Kushtia', 'Khulna'),
  District('Magura', 'Khulna'),
  District('Meherpur', 'Khulna'),
  District('Narail', 'Khulna'),
  District('Satkhira', 'Khulna'),
  // Barishal
  District('Barguna', 'Barishal'),
  District('Barishal', 'Barishal'),
  District('Bhola', 'Barishal'),
  District('Jhalokati', 'Barishal'),
  District('Patuakhali', 'Barishal'),
  District('Pirojpur', 'Barishal'),
  // Sylhet
  District('Habiganj', 'Sylhet'),
  District('Moulvibazar', 'Sylhet'),
  District('Sunamganj', 'Sylhet'),
  District('Sylhet', 'Sylhet'),
  // Rangpur
  District('Dinajpur', 'Rangpur'),
  District('Gaibandha', 'Rangpur'),
  District('Kurigram', 'Rangpur'),
  District('Lalmonirhat', 'Rangpur'),
  District('Nilphamari', 'Rangpur'),
  District('Panchagarh', 'Rangpur'),
  District('Rangpur', 'Rangpur'),
  District('Thakurgaon', 'Rangpur'),
  // Mymensingh
  District('Jamalpur', 'Mymensingh'),
  District('Mymensingh', 'Mymensingh'),
  District('Netrokona', 'Mymensingh'),
  District('Sherpur', 'Mymensingh'),
];

/// Old and common alternative spellings, mapped to the official name.
/// Most are the pre-2018 English spellings still printed on older signs
/// and forms. "Savar" is here because early versions of this app used it
/// as a default; it is an upazila of Dhaka district.
const _aliases = <String, String>{
  'bogra': 'Bogura',
  'chittagong': 'Chattogram',
  'comilla': 'Cumilla',
  'barisal': 'Barishal',
  'jessore': 'Jashore',
  'nawabganj': 'Chapai Nawabganj',
  'laxmipur': 'Lakshmipur',
  'maulvibazar': 'Moulvibazar',
  'netrakona': 'Netrokona',
  'jhalakathi': 'Jhalokati',
  'jhalokathi': 'Jhalokati',
  'khagrachari': 'Khagrachhari',
  'narshingdi': 'Narsingdi',
  'hobiganj': 'Habiganj',
  'jhenaidaha': 'Jhenaidah',
  'coxbazar': "Cox's Bazar",
  'savar': 'Dhaka',
};

String _key(String s) => s.toLowerCase().replaceAll(RegExp(r'[^a-z]'), '');

final Map<String, String> _byKey = {
  for (final d in districts) _key(d.name): d.name,
  ..._aliases,
};

/// The official name for [input], or null if it is not a district.
///
/// Ignores case, spaces and punctuation, and accepts the common old
/// spellings, so "bogra", "BOGURA" and " Bogura " all give "Bogura".
String? normalizeDistrict(String input) => _byKey[_key(input)];

bool isKnownDistrict(String input) => normalizeDistrict(input) != null;

String? divisionOf(String name) {
  for (final d in districts) {
    if (d.name == name) return d.division;
  }
  return null;
}

File: backend/districts.py
"""
The 64 districts of Bangladesh, matching lib/core/districts.dart.

Reports are grouped by district, so the server only accepts official
names. This fixes two problems at once: misspellings that would split one
district's reports across several names, and invented district names
that would let anyone create outbreak buckets at will.

It also replaces the old str.title() normalisation, which turned
"Cox's Bazar" into "Cox'S Bazar".

Change this file and the Dart one together.
"""

import re

DISTRICTS = [
    "Dhaka",
    "Faridpur",
    "Gazipur",
    "Gopalganj",
    "Kishoreganj",
    "Madaripur",
    "Manikganj",
    "Munshiganj",
    "Narayanganj",
    "Narsingdi",
    "Rajbari",
    "Shariatpur",
    "Tangail",
    "Bandarban",
    "Brahmanbaria",
    "Chandpur",
    "Chattogram",
    "Cox's Bazar",
    "Cumilla",
    "Feni",
    "Khagrachhari",
    "Lakshmipur",
    "Noakhali",
    "Rangamati",
    "Bogura",
    "Chapai Nawabganj",
    "Joypurhat",
    "Naogaon",
    "Natore",
    "Pabna",
    "Rajshahi",
    "Sirajganj",
    "Bagerhat",
    "Chuadanga",
    "Jashore",
    "Jhenaidah",
    "Khulna",
    "Kushtia",
    "Magura",
    "Meherpur",
    "Narail",
    "Satkhira",
    "Barguna",
    "Barishal",
    "Bhola",
    "Jhalokati",
    "Patuakhali",
    "Pirojpur",
    "Habiganj",
    "Moulvibazar",
    "Sunamganj",
    "Sylhet",
    "Dinajpur",
    "Gaibandha",
    "Kurigram",
    "Lalmonirhat",
    "Nilphamari",
    "Panchagarh",
    "Rangpur",
    "Thakurgaon",
    "Jamalpur",
    "Mymensingh",
    "Netrokona",
    "Sherpur",
]

# Old spellings still seen on forms and signs, and "Savar", an upazila
# early versions of the app used as a default district.
ALIASES = {
    "bogra": "Bogura",
    "chittagong": "Chattogram",
    "comilla": "Cumilla",
    "barisal": "Barishal",
    "jessore": "Jashore",
    "nawabganj": "Chapai Nawabganj",
    "laxmipur": "Lakshmipur",
    "maulvibazar": "Moulvibazar",
    "netrakona": "Netrokona",
    "jhalakathi": "Jhalokati",
    "jhalokathi": "Jhalokati",
    "khagrachari": "Khagrachhari",
    "narshingdi": "Narsingdi",
    "hobiganj": "Habiganj",
    "jhenaidaha": "Jhenaidah",
    "coxbazar": "Cox's Bazar",
    "savar": "Dhaka",
}


def _key(s: str) -> str:
    return re.sub(r"[^a-z]", "", s.lower())


_BY_KEY = {_key(d): d for d in DISTRICTS}
_BY_KEY.update(ALIASES)


def normalize_district(value: str):
    """The official name for value, or None if it is not a district."""
    return _BY_KEY.get(_key(value or ""))
